### PR TITLE
Catch errors parsing build.yaml

### PR DESCRIPTION
**Pull request by natebosch**
Towards #1426 

Instead of a stack trace the build will fail with an error message.

### CODE REVIEW

File: build_runner/lib/src/build_script_generate/build_script_generate.dart [hunk 1]
```diff
@@ -50,8 +50,14 @@ Future<Iterable<Expression>> _findBuilderApplications(String configKey) async {
     if (buildConfigOverrides.containsKey(package.name)) {
       return buildConfigOverrides[package.name];
     }
-    return BuildConfig.fromBuildConfigDir(
-        package.name, package.dependencies.map((n) => n.name), package.path);
+    try {
+      return await BuildConfig.fromBuildConfigDir(
+          package.name, package.dependencies.map((n) => n.name), package.path);
+    } catch (e) {
```
Comment (kevmoo) [inline build_runner/lib/src/build_script_generate/build_script_generate.dart]: could this really be anything? catch alls make me nervous

File: build_runner/lib/src/generate/build_impl.dart [hunk 2]
```diff
@@ -71,13 +71,24 @@ Future<BuildResult> build(
...
       reader: reader,
       writer: writer,
       onLog: onLog);
+  TargetGraph targetGraph;
+  try {
+    targetGraph = await TargetGraph.forPackageGraph(packageGraph,
+        overrideBuildConfig: overrideBuildConfig);
+  } on BuildConfigParseException catch (e) {
+    // Logging is not set up until the BuildOptions is created
+    environment.onLog(new LogRecord(
```
Comment (kevmoo) [inline build_runner/lib/src/generate/build_impl.dart]: no way to capture the stack?

File: build_runner/lib/src/generate/watch_impl.dart [hunk 3]
```diff
@@ -68,14 +68,25 @@ Future<ServeHandler> watch(
...
       writer: writer,
       directoryWatcherFactory: directoryWatcherFactory,
       onLog: onLog);
+  TargetGraph targetGraph;
+  try {
+    targetGraph = await TargetGraph.forPackageGraph(packageGraph,
+        overrideBuildConfig: overrideBuildConfig);
+  } on BuildConfigParseException catch (e) {
+    // Logging is not set up until the BuildOptions is created
+    environment.onLog(new LogRecord(
```
Comment (kevmoo) [inline build_runner/lib/src/generate/watch_impl.dart]: must catch all?
no way to capture the stack?

File: build_runner/lib/src/package_graph/target_graph.dart [hunk 4]
```diff
@@ -52,6 +52,17 @@ Future<BuildConfig> _packageBuildConfig(PackageNode package) async {
   if (package.path == null) {
     return new BuildConfig.useDefault(package.name, dependencyNames);
   }
-  return BuildConfig.fromBuildConfigDir(
-      package.name, dependencyNames, package.path);
+  try {
+    return await BuildConfig.fromBuildConfigDir(
+        package.name, dependencyNames, package.path);
+  } catch (e) {
+    throw new BuildConfigParseException(package.name, e);
```
Comment (kevmoo) [inline build_runner/lib/src/package_graph/target_graph.dart]: ditto...

File: build_runner/lib/src/build_script_generate/build_script_generate.dart [hunk 5]
```diff
@@ -50,8 +50,14 @@ Future<Iterable<Expression>> _findBuilderApplications(String configKey) async {
...
+      // During the build an error will be logged
```
Comment (grouma) [inline build_runner/lib/src/build_script_generate/build_script_generate.dart]: [nit] end comments with a period (here and below): https://www.dartlang.org/guides/language/effective-dart/documentation#do-format-comments-like-sentences

File: build_runner/lib/src/generate/build_impl.dart (hunk 2 above)
Comment (natebosch) [reply build_runner/lib/src/generate/build_impl.dart]: We don't care about the stack here.

File: build_runner/lib/src/generate/watch_impl.dart (hunk 3 above)
Comment (natebosch) [reply build_runner/lib/src/generate/watch_impl.dart]: `on BuildConfigParseException`

Comment (kevmoo) [reply build_runner/lib/src/generate/watch_impl.dart]: ah...sorry...

File: build_runner/lib/src/generate/build_impl.dart (hunk 2 above)
Comment (kevmoo) [reply build_runner/lib/src/generate/build_impl.dart]: ack

File: build_runner/lib/src/build_script_generate/build_script_generate.dart (hunk 1 above)
Comment (natebosch) [reply build_runner/lib/src/build_script_generate/build_script_generate.dart]: catching `ArgumentError` now.

File: build_runner/lib/src/build_script_generate/build_script_generate.dart (hunk 5 above)
Comment (natebosch) [reply build_runner/lib/src/build_script_generate/build_script_generate.dart]: Done

**Comment by natebosch** [comment]
Added a new exception `CannotBuildException` to indicate that something in the environment is broken and builds are not possible. Today it means that some dependency has a bad `build.yaml` file.

Exposing this exception since users of `build` or `watch` will need to deal with it. Users of `run` will get an appropriate exit code.

File: build_runner/lib/src/generate/watch_impl.dart [hunk 6]
```diff
@@ -62,7 +62,6 @@ Future<ServeHandler> watch(
 }) async {
   builderConfigOverrides ??= const {};
   packageGraph ??= new PackageGraph.forThisPackage();
-
```
Comment (grouma) [inline build_runner/lib/src/generate/watch_impl.dart]: Unnecessary change?

Comment (natebosch) [reply build_runner/lib/src/generate/watch_impl.dart]: yup, had previously needed a change in this file before someone went and gave me merge conflicts 😝 

reverted
